servers/src/main/java/tachyon/master/lineage/LineageMaster.java

```diff
@@ -0,0 +1,421 @@
+    for (TachyonURI inputFile : inputFiles) {
+      long fileId;
+      fileId = mFileSystemMaster.getFileId(inputFile);
+      inputTachyonFiles.add(new TachyonFile(fileId));
+    }
+    // create output files
+    List<LineageFile> outputTachyonFiles = Lists.newArrayList();
+    for (TachyonURI outputFile : outputFiles) {
+      long fileId;
+      // create the file initialized with block size 1 as placeholder
+      fileId = mFileSystemMaster.createFile(outputFile, 1, true);
+      outputTachyonFiles.add(new LineageFile(fileId));
+    }
+
+    LOG.info("Create lineage of input:" + inputTachyonFiles + ", output:" + outputTachyonFiles
+        + ", job:" + job);
+    long lineageId = mLineageStore.createLineage(inputTachyonFiles, outputTachyonFiles, job);
+
+    writeJournalEntry(mLineageIdGenerator.toJournalEntry());
+    writeJournalEntry(mLineageStore.getLineage(lineageId).toJournalEntry());
+    flushJournal();
+    return lineageId;
+  }
+
+  public boolean deleteLineage(long lineageId, boolean cascade)
+      throws LineageDoesNotExistException, LineageDeletionException {
+    Lineage lineage = mLineageStore.getLineage(lineageId);
+    if (lineage == null) {
+      throw new LineageDoesNotExistException(
+          "the lineage " + lineageId + " to delete does not exist");
+    }
+
+    // there should not be child lineage if not cascade
+    if (!cascade && !mLineageStore.getChildren(lineage).isEmpty()) {
+      throw new LineageDeletionException(
+          "the lineage " + lineageId + " to delete has children lineages");
+    }
+
+    LOG.info("Delete lineage " + lineageId);
+    mLineageStore.deleteLineage(lineageId);
+    return true;
+  }
+
+  /**
+   * Reinitializes the file when the file is lost or not completed.
+   *
+   * @param path the path to the file
+   * @param blockSizeBytes the block size
+   * @param ttl the TTL
+   * @return the id of the reinitialized file when the file is lost or not completed, -1 otherwise
+   * @throws InvalidPathException the file path is invalid
+   */
+  public long recreateFile(String path, long blockSizeBytes, long ttl)
+      throws InvalidPathException, LineageDoesNotExistException {
+    long fileId = mFileSystemMaster.getFileId(new TachyonURI(path));
+    LineageFileState state = mLineageStore.getLineageFileState(fileId);
+    if (state == LineageFileState.CREATED || state == LineageFileState.LOST) {
+      LOG.info("Recreate the file " + path + " with block size of " + blockSizeBytes + " bytes");
+      return mFileSystemMaster.reinitializeFile(new TachyonURI(path), blockSizeBytes, ttl);
+    }
+    return -1;
+  }
+
+  public void asyncCompleteFile(long fileId)
+      throws FileDoesNotExistException, BlockInfoException {
+    LOG.info("Async complete file " + fileId);
+    mLineageStore.completeFile(fileId);
+    // complete file in Tachyon.
+    mFileSystemMaster.completeFile(fileId);
+    writeJournalEntry(new AsyncCompleteFileEntry(fileId));
+    flushJournal();
```

Comment:
synchronize on both operation and journaling? If the first completeFile succeeds and the file is aborted before the second completeFile, what's the assumed behavior?